backend/data_preparation/crawler/fire_crawler.py

```diff
@@ -0,0 +1,177 @@
+            # to change "current_year" to the real year as an integer
+            # ignore years that before the start year
+            # filter out links that is too old
+            if true_year < start_year:
+                continue
+            for state in self.states:
+                # try except for network errors
+                list_of_state_fires = self.get_url(f"{self.baseDir}{year_node}/{state}",
+                                                   f"{state} State of year {year_node}",
+                                                   "extract_all_fires()", 10)
+                # crawl the all fires in this state
+                re_formula = r'<A .*?>(.*?)</A>'
+                # This re_formula is different from the one used
+                firenames = re.findall(re_formula, list_of_state_fires, re.S | re.M)
+                fires.extend((map(lambda f: (true_year, state, f),firenames)))
+        # to delete exceptions: "[To Parent Directory]" "whatever.zip"
+        # ActivePerim is an useless link which contains metadata over the year, existing in links before 2016 Skip it
+        for year, state, firename in fires[:]:
+            if "." in firename or "[To Parent Directory]" == firename or firename == "ActivePerim":
+                fires.remove((year, state, firename))
+        return fires
+
+    @staticmethod
+    def generate_url_from_tuple(tuple_year: int, tuple_state: str, tuple_firename: str, current_year: int) -> str:
+        """
+        Take year, state, and url name and convert the entry to the url
+        :param tuple_year: int
+        :param tuple_state: str
+        :param tuple_firename: str
+        :param current_year: int
+        :return: str
+        """
+        baseDir = 'https://rmgsc.cr.usgs.gov/outgoing/GeoMAC/'
+        year_str = "current_year" if tuple_year == current_year else str(tuple_year)
+        url = f"{baseDir}{year_str}_fire_data/{tuple_state}/{tuple_firename}"
+        return url
+
+    def crawl(self, url_to_crawl):
+        """
+        Takes the url string of one fire and start to crawl all files
+        :param url_to_crawl: str
+        :return: None
+        """
+        used_folder_names = set()
+        # grab all firenames in the page
+        # for loop for runtime error
+        fire_page = self.get_url(url_to_crawl, url_to_crawl, "crawl()", 10)
+        res = r'<A .*?>(ca_.+?)</A>'
+        filenames = re.findall(res, fire_page, re.S | re.M)
+        # download useful files
+        for file in filenames:
+            if not ".zip" == file[-4:]:
+                # if it is a zip don't download
+                folder_name = file.split(".")[0]
+                if folder_name not in used_folder_names:
+                    used_folder_names.add(folder_name)
+                    os.makedirs(os.path.join(FIRE_DATA_DIR,folder_name))
+                out_path = os.path.join(FIRE_DATA_DIR, folder_name)
+                for attempts in range(10):
```

Comment:
2 questions:
1. can we use decorator for making such attempts?
2. why 10 times? why statically 10 times?